src/tooling/piral-cli-webpack5/src/enhancers/pilet-webpack-config-enhancer/index.ts

```diff
@@ -0,0 +1,104 @@
+import { BannerPlugin, DefinePlugin, WebpackPluginInstance } from 'webpack';
+import { setEnvironment, getDefineVariables, getVariables } from './helpers';
+
+export interface PiletWebpackConfigEnhancerOptions {
+  /**
+   * The name of the pilet.
+   */
+  name: string;
+  /**
+   * The version of the pilet.
+   */
+  version: string;
+  /**
+   * The name of the Piral instance / app shell.
+   */
+  piral: string;
+  /**
+   * The schema version. By default, v1 is used.
+   */
+  schema?: 'v0' | 'v1' | 'none';
+  /**
+   * The shared dependencies. By default, these are read from the
+   * Piral instance.
+   */
+  externals?: Array<string>;
+  /**
+   * Additional environment variables to define.
+   */
+  variables?: Record<string, string>;
+}
+
+function getExternals(piral: string) {
+  const shellPkg = require(`${piral}/package.json`);
+  const piralExternals = shellPkg.pilets?.externals ?? [];
+  return [
+    ...piralExternals,
+    '@dbeining/react-atom',
+    '@libre/atom',
+    'history',
+    'react',
+    'react-dom',
+    'react-router',
+    'react-router-dom',
+    'tslib',
+    'path-to-regexp',
+  ];
+}
+
+export const piletWebpackConfigEnhancer = (options: PiletWebpackConfigEnhancerOptions) => (compilerOptions) => {
+  const environment = process.env.NODE_ENV || 'development';
+  const { name, version, piral, externals = getExternals(piral), schema } = options;
+  const shortName = name.replace(/\W/gi, '');
+  const jsonpFunction = `pr_${shortName}`;
+  const variables = {
+    ...getVariables(name, version, environment),
+    ...options.variables,
+  };
+  const plugins: WebpackPluginInstance[] = [new DefinePlugin(getDefineVariables(variables))];
+
+  if (schema !== 'none') {
+    const bannerSuffix = schema ? `1(${jsonpFunction})` : `0`;
+
+    plugins.push(
+      new BannerPlugin({
+        banner: `//@pilet v:${bannerSuffix}`,
+        entryOnly: true,
+        include: /\.js$/,
+        raw: true,
+      }),
+    );
+  }
+
+  // FIXME: this is a really bad idea, making finding the source of configuration nearly impossible
+  setEnvironment(variables);
```

Comment:
Is this still needed?

Reply:
well, that's why it's a bad idea – the information passing is implicit and I have no clue who is relying on this.
I have no idea whether it's needed or not since this is not my code (copied from the plugin).

Reply:
> I have no idea whether it's needed or not since this is not my code (copied from the plugin).

Has this indeed anything to do with that specific part (that argument could be brought for any functionality from another dependency)? In any case that is the problem with environment variables, however, as Webpack 5 changed the use of environment variables I'm not sure its needed any more.

A bit of testing would reveal it.

I think the `NODE_ENV` may be implicitly set by Webpack. The rest I don't actually know, but I assume it may also not matter any more. We can leave it as is (would not hurt much, too).